mlir/lib/Conversion/GPUToMIGraphX/GPUToMIGraphX.cpp

```diff
@@ -0,0 +1,122 @@
+//===- GPUToMIGraphX.cpp - Lowering GPU to MIGraphX Dialect -----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// These rewriters lower from the GPU to the MIGraphX dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/GPUToMIGraphX/GPUToMIGraphX.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/GPU/GPUDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+
+using namespace mlir;
+
+namespace {
+
+class FuncToCOBJPattern : public OpConversionPattern<CallOp> {
+  using OpConversionPattern<CallOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(CallOp op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op->getLoc();
+    auto results = op->getResults();
+    auto resultType = results[0].getType().template cast<MemRefType>();
+
+    // Insert alloc for result buffer
+    rewriter.setInsertionPoint(op);
+    auto resultAlloc = rewriter.create<memref::AllocOp>(loc, resultType);
+    
+    auto fnAttr = op->getAttrOfType<FlatSymbolRefAttr>("callee");
+    SmallVector<Value, 8> mrOperands(op.getOperands());
+    SmallVector<Value, 8> cobjArgs;
+    mrOperands.push_back(resultAlloc);
+
+    SmallVector<IntegerAttr, 5> globalSizeAttr;
+    SmallVector<IntegerAttr, 5> localSizeAttr;
+    SymbolRefAttr kernelRefAttr;
+
+    auto fusedFuncOp =
+      op->getParentOfType<ModuleOp>().lookupSymbol<FuncOp>(fnAttr.getValue());
+
```

Comment:
Also check the kernelFunc is found and may have already been processed by other calls.

Reply:
Nevermind. This does not change the kernelFunc just the calls so it is correct.